### PR TITLE
Admin UI: user management

### PR DESCRIPTION
**Pull request by jnu**
Admin UI for creating, listing, editing, and deleting users.

in this pr
 - [x] create new API
 - [x] edit user UI
 - [x] list all users UI
 - [x] merge api changes into #77 
 - [x] forgot password UI flow
 - [x] verification email UI handler
 - [x] email send handler for user registration / forgot password / verification
 - [x] tests for fastapi-users routes
 - [x] tests for graphql schema changes
 - [x] Accessibility questions about Forms
 - [x] UI polish
 - [x] UI tests

### CODE REVIEW

**Comment by jnu** [comment]
screens to get a gist of the UI changes:

==admin screens==

User list:
image

Create user [clicking save redirects to the Edit user page below. I don't love the UI, but it fits most easily with how the fastapi-users API was designed (e.g., you can't assign admin privileges when you create a user, only when editing)]:
image

Edit a user:
image

Assign to a team:
image

Deactivation modal:
image

Editing inactive user / reactivating:
image

==general user screens==

Forgot password on login:
image

Email confirmation:
image

Reset password screen [the token in the URL authorizes this request, then redirects to login on success]:
image

Error if the token is invalid:
image

File: api/app.py
Comment (seerocode) [inline api/app.py:127]: PR has been merged so this can be changed now: https://github.com/frankie567/fastapi-users/pull/650#issuecomment-873448596

File: api/mailer.py
Comment (seerocode) [inline api/mailer.py:64]: Do you have a screenshot of what this looks like when it's converted from HTML to ascii text?

File: api/queries.py
Comment (seerocode) [inline api/queries.py:33]: The users query will now return all users whether they were deleted or not and an "active" field to check their status in the User object?

File: api/queries.py
Comment (seerocode) [inline api/queries.py:159]: We could probably add this delete check to the database class for consistency.

File: api/settings.py
Comment (seerocode) [inline api/settings.py:11]: Do we need to handle this port at all in the Docker container?

File: client/src/components/Admin/CreateUser.tsx
Comment (seerocode) [inline client/src/components/Admin/CreateUser.tsx]: Can we check the accessibility on these forms? I had used a native HTML form for the data input page because there were accessibility issues raised in devtools when I tried to use a AntD form. Hopefully it's been fixed.

File: client/src/components/Admin/UserList.tsx
Comment (seerocode) [inline client/src/components/Admin/UserList.tsx]: We can use the `ErrorFallback` component here. It's the AntD Error component.

File: client/src/components/Admin/UserList.tsx
Comment (seerocode) [inline client/src/components/Admin/UserList.tsx]: This is saying these array items are keys of `GetUserList_users`?

File: client/src/components/Admin/UserList.tsx
Comment (seerocode) [inline client/src/components/Admin/UserList.tsx]: Can `c` be renamed to something more descriptive? I think I did this too in some places and ended up renaming in subsequent PRs.

File: client/src/components/VerifyAccount.tsx
Comment (seerocode) [inline client/src/components/VerifyAccount.tsx]: This is a full-page spinner? If so, maybe we do the skeleton loading instead of a spinner? Fine for now.

File: client/src/router/routes.tsx
Comment (seerocode) [inline client/src/router/routes.tsx:93]: Small nit! The casing is different here than the rest of the breadcrumbs.

File: client/src/services/account.ts
Comment (seerocode) [inline client/src/services/account.ts:17]: Should we have an interface for roles?

File: client/src/services/account.ts
Comment (seerocode) [inline client/src/services/account.ts]: Might be good to move this to a global error handler in the future.

File: api/settings.py
Comment (jnu) [reply api/settings.py:11]: good question! this is an outbound connection (our web app sends data to the specified SMTP server, which mails the message), which should be permitted by default.

i added a chore to test verify that this all works with a real smtp server, in part because the rendered HTML also sometimes looks funky:
https://app.clubhouse.io/stanford-computational-policy-lab/story/307/test-sending-real-emails-via-smtp

File: client/src/components/VerifyAccount.tsx
Comment (jnu) [reply client/src/components/VerifyAccount.tsx]: what do you mean by skeleton loading?

File: client/src/services/account.ts
Comment (jnu) [reply client/src/services/account.ts:17]: what do you mean by interface? the type at this point is a string (just the role ID).

File: client/src/services/account.ts
Comment (jnu) [reply client/src/services/account.ts]: the error responses are somewhat different based on the endpoint, but there may be some common stuff we could factor out

File: api/mailer.py
Comment (jnu) [reply api/mailer.py:64]: it basically just adds paragraph breaks:

image

for context, email is an old protocol and by default emails are plain text, without formatting. emails these days are generally sent as "multipart" messages that specify both a text/plain version for backwards-compatibility and a text/html version for nice formatting. nobody ever sees the text/plain version of an email unless they explicitly turn off html formatting in their email client.

File: api/app.py
Comment (jnu) [reply api/app.py:127]: unfortunately we should wait for a new tagged release version before removing. you _can_ specify a git sha in requirements.txt, but you run the risk of pulling intermediate / broken changes, so it's better to wait for the maintainers to ship a release. since the workaround is small i don't mind waiting.

File: api/queries.py
Comment (jnu) [reply api/queries.py:33]: that's correct. the `users` query is new so worth noting i didn't change any behavior there. the `user` (singular) query I modified to return inactive users to support the "reactivate" button.

File: api/queries.py
Comment (jnu) [reply api/queries.py:159]: the non-classmethods actually look like the norm in this file, so i'd say it's consistent! :p 

i agree we should pick a strategy and stick with it. worth nothing we have to rewrite all these queries to support https://app.clubhouse.io/stanford-computational-policy-lab/story/302/use-async-db-calls so i think it's ok to punt for now.

File: client/src/components/Admin/CreateUser.tsx
Comment (jnu) [reply client/src/components/Admin/CreateUser.tsx]: interesting, i will check accessibility. i realized as i was writing this that the issues you were having with the submit behavior were likely due to mixing the native html form with the Input elements from antd. antd also has a well-developed system of form validation (and UI layout, including labels and error messages) that we'd miss out on if we dropped their Form component.

i'll investigate the accessibility issues -- i'd much rather find a workaround and/or fix the issues upstream (as i did for the fastapi-users) than not be able to use their components.

File: client/src/components/Admin/UserList.tsx
Comment (jnu) [reply client/src/components/Admin/UserList.tsx]: yep! i specified this so the `sorter` function could be typechecked properly. typescript will also complain if you put items in that list that _aren't_ keys of that users object.

File: api/app.py
Comment (seerocode) [reply api/app.py:127]: Probably better to include as a story then since it may be easy to forget.

File: client/src/components/Admin/UserList.tsx
Comment (jnu) [reply client/src/components/Admin/UserList.tsx]: ah good to know, thanks! fwiw i saw some TODOs about loading/errors in other files that i essentially copied here. will update.

Comment (seerocode) [reply client/src/components/Admin/UserList.tsx]: Yeah :/ I've been cleaning them up as I go. It would be better to have like an error wrapper but there are issues with async stuff there.

File: client/src/components/VerifyAccount.tsx
Comment (seerocode) [reply client/src/components/VerifyAccount.tsx]: https://ant.design/components/skeleton/

File: client/src/services/account.ts
Comment (seerocode) [reply client/src/services/account.ts:17]: There was another place where roles was defined as either admin or something else. I can't find where it was now.

Comment (jnu) [reply client/src/services/account.ts:17]: ah, you mean a string enum? like `type Role = 'admin';`?

for this part of the UI i wanted to make it fairly general -- we just request a list of all defined roles from the server (in case we add more later) and the UI uses their IDs (not their names) to toggle them on and off. so this "edit user" stuff doesn't know anything in particular about what any of the roles are. the IDs are just UUIDs, which are arbitrary.

in contrast, for the `auth` part of the UI it makes sense to check the `admin` role name explicitly to check if the user has that particular role.

File: client/src/components/VerifyAccount.tsx
Comment (jnu) [reply client/src/components/VerifyAccount.tsx]: ah i see, maybe i'm not sure what the skeleton would indicate though? there's not actually any UI on this page that's getting loaded -- in the normal case all this page does is redirect the user somewhere else. The flow is:

1) User gets an email requesting they verify their email by clicking the given link
2) User navigates to that link
3) That link sends the token to the backend for verification
4) Assuming all goes well they get redirected pretty much immediately

Comment (seerocode) [reply client/src/components/VerifyAccount.tsx]: Got it! The current spinner makes sense here then.

File: api/app.py
Comment (jnu) [reply api/app.py:127]: great idea, tracked here:
https://app.clubhouse.io/stanford-computational-policy-lab/story/308/remove-hack-for-fastapi-users-delete-handler

**Review by seerocode — APPROVED** [review]

File: client/src/components/Admin/UserList.tsx
Comment (jnu) [reply client/src/components/Admin/UserList.tsx]: sure, i'm always super lazy about iteration variable names :(

**Pull request merged**
